### PR TITLE
Use VersionRange to Refactor Versioned SQL Stores

### PR DESCRIPTION
**Pull request by wsugarman**
## Description
- Add new type `VersionRange` to re-use SQL stores for multiple versions

## Related issues
N/A

## Testing
Test locally

### CODE REVIEW

**Review by pengchen0692 — APPROVED** [review]
I'm thinking if we could have better naming pattern for different versions of stores, current naming is confusing. 
Anyway, we could solve that problem later.

File: src/Microsoft.Health.Dicom.SqlServer/Features/Schema/VersionRange.cs [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+            Max = max;
+        }
+
+        public bool Contains(SchemaVersion version)
+        {
+            EnsureArg.EnumIsDefined(version, nameof(version));
+            return Min <= version && version <= Max;
+        }
+
+        public void Deconstruct(out SchemaVersion min, out SchemaVersion max)
```
[the remark this answers is not in the document]
Comment (wsugarman) [reply src/Microsoft.Health.Dicom.SqlServer/Features/Schema/VersionRange.cs:45]: I think that would be confusing. The deconstruct is meant for use when interacting with tuples

File: src/Microsoft.Health.Dicom.SqlServer/Features/Schema/VersionRange.cs [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+        }
+
+        public override bool Equals(object obj)
+            => obj is VersionRange other && Equals(other);
+
+        public bool Equals(VersionRange other)
+            => Min == other.Min && Max == other.Max;
+
+        public override int GetHashCode()
+            => Min.GetHashCode() ^ Max.GetHashCode();
```
[the remark this answers is not in the document]
Comment (wsugarman) [reply src/Microsoft.Health.Dicom.SqlServer/Features/Schema/VersionRange.cs:58]: Yeah, I can use it. It's perf is not really that great compared to what the dotnet team uses internally (and certainly compared to a simple XOR). But I'm not planning on this being called... really at all lol. It's more for completeness.

They talk about it briefly here: https://github.com/dotnet/runtime/issues/13377

File: src/Microsoft.Health.Dicom.SqlServer/Features/Common/SqlStoreFactory.cs [hunk 3]
```diff
@@ -28,10 +28,10 @@ internal class SqlStoreFactory<TVersionedStore, TStore> : IStoreFactory<TStore>
...
         public SqlStoreFactory(ISchemaVersionResolver schemaResolver, IEnumerable<TVersionedStore> versionedStores)
         {
-            EnsureArg.IsNotNull(schemaResolver, nameof(schemaResolver));
-            EnsureArg.IsNotNull(versionedStores, nameof(versionedStores));
-            _schemaResolver = schemaResolver;
-            _versionedStores = versionedStores.ToDictionary(x => x.Version);
+            _schemaResolver = EnsureArg.IsNotNull(schemaResolver, nameof(schemaResolver));
+            _versionedStores = EnsureArg.IsNotNull(versionedStores, nameof(versionedStores))
+                .SelectMany(x => x.SupportedVersions.Select(v => (Version: v, Store: x)))
+                .ToDictionary(x => x.Version, x => x.Store);
```
[the remark this answers is not in the document]
Comment (wsugarman) [reply src/Microsoft.Health.Dicom.SqlServer/Features/Common/SqlStoreFactory.cs:34]: I can use a different variable for both, but they are the same type in each delegate so I hesitate to use different variables.

[the remark this answers is not in the document]
Comment (pengchen0692) [reply src/Microsoft.Health.Dicom.SqlServer/Features/Common/SqlStoreFactory.cs]: Sorry, I mean the 2nd and 3rd `x`.  `x` is totally used for 3 times in the expression. 
Consider something like: 
`.SelectMany(store => store.SupportedVersions.Select(version => (Version: version, Store: store)))
                .ToDictionary(pair => pair.Version, x => pair.Store);`

**Pull request closed without merging**